packages/dds/map/src/mapKernel.ts

```diff
@@ -753,20 +1135,60 @@ export class MapKernel {
 		} else {
 			this.pendingKeys.set(op.key, [pendingMessageId]);
 		}
+
 		return pendingMessageId;
 	}
 
+	private incrementPendingDeleteCount(key: string): void {
+		const count = this.pendingDeleteTracker.get(key) ?? 0;
+		this.pendingDeleteTracker.set(key, count + 1);
+	}
+
+	private decrementPendingDeleteCount(key: string): void {
+		const count = this.pendingDeleteTracker.get(key) ?? 0;
+		this.pendingDeleteTracker.set(key, count - 1);
+		if (count <= 1) {
+			this.pendingDeleteTracker.delete(key);
+		}
+	}
+
+	/**
+	 * Update the message id's associated with the op's key
+	 * @param op - The map key message
+	 * @param pendingMessageId - The associated message id to be handled
+	 */
+	private updatePendingSetIds(op: IMapKeyOperation, pendingMessageId: number): void {
+		// Store the messageId, for the creation order of unack'd keys
+		if (op.type === "set") {
+			if (!this.pendingSetTracker.has(op.key)) {
+				this.pendingSetTracker.set(op.key, [pendingMessageId]);
+				this.localKeysIndexTracker.set(op.key, pendingMessageId);
+			} else {
+				this.pendingSetTracker.get(op.key)?.push(pendingMessageId);
+			}
+		} else {
+			if (this.pendingSetTracker.has(op.key)) {
+				const index = this.pendingSetTracker.get(op.key)?.[0];
+				this.pendingSetTracker.delete(op.key);
+				this.localKeysIndexTracker.delete(index as number);
+			}
+		}
+	}
+
 	/**
 	 * Submit a map key message to remote clients.
```

Comment:
Similar feedback to before on casting--we should probably be doing some runtime validation here.